pkg/auth/github.go

```diff
@@ -0,0 +1,172 @@
+const githubRequredGrantType = "urn:ietf:params:oauth:grant-type:device_code"
+
+// It appears we need `repo` scope, which is VERY permissive.
+// We need to be able to push a deploy key and merge commits. No other scopes matched.
+// Available scopes: https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps
+const githubOAuthScope = "repo"
+
+type githubAuthResponse struct {
+	AccessToken string `json:"access_token"`
+	Error       string `json:"error"`
+}
+
+// doAuthRequest is used to poll for the status of the device flow.
+func doAuthRequest(client *http.Client, deviceCode string) (string, error) {
+	query := url.Values.Encode(map[string][]string{
+		"client_id":   {WeGOGithubClientID},
+		"device_code": {deviceCode},
+		"grant_type":  {githubRequredGrantType},
+	})
+	url := fmt.Sprintf(accessTokenUrl, query)
+
+	req, err := http.NewRequest("POST", url, nil)
+	if err != nil {
+		return "", fmt.Errorf("could not create auth request: %w", err)
+	}
+
+	b, err := doRequest(req, client)
+	if err != nil {
+		return "", fmt.Errorf("error doing auth request: %w", err)
+	}
+
+	p := githubAuthResponse{}
+
+	if err := json.Unmarshal(b, &p); err != nil {
+		return "", fmt.Errorf("err marshaling request body: %w", err)
+	}
+
+	if p.Error == "authorization_pending" {
+		// This is expect until the user completes the auth flow.
+		return "", ErrAuthPending
+	}
+
+	if p.AccessToken != "" {
+		return p.AccessToken, nil
+	}
+
+	// Note p.Error is a string here
+	return "", fmt.Errorf("error doing auth request: %s", p.Error)
+}
+
+// NewGithubDeviceFlowHandler returns a function which will initiate the Github Device Flow for the CLI.
+func NewGithubDeviceFlowHandler(client *http.Client) BlockingCLIAuthHandler {
+	return func(ctx context.Context, w io.Writer) (string, error) {
+		codeRes, err := doCodeRequest(client, githubOAuthScope)
+		if err != nil {
+			return "", fmt.Errorf("could not do code request: %w", err)
+		}
+
+		fmt.Fprintln(w)
+		fmt.Fprintf(w, "Visit this URL to authenticate with Github:\n\n")
+		fmt.Fprintf(w, "%s\n\n", codeRes.VerficationURI)
+		fmt.Fprintf(w, "Type the following code into the page at the URL above: %s\n\n", codeRes.UserCode)
+		fmt.Fprintf(w, "Waiting for authentication flow completion...\n\n")
+
+		// GH complains if you retry RIGHT at the given interval.
+		// We will get a `slow_down` error from the backend without the one second padding.
+		retryInterval := time.Duration(codeRes.Interval+1) * time.Second
+
+		ticker := time.NewTicker(retryInterval)
+
+		for range ticker.C {
+			authToken, err := doAuthRequest(client, codeRes.DeviceCode)
+			if err != nil {
+				if err == ErrAuthPending {
+					// This is expected while the user goes to the webpage.
+					continue
+				}
+
+				// An unexpected error happened, return it.
+				ticker.Stop()
+				return "", err
+			}
+
+			ticker.Stop()
+			fmt.Fprintf(w, "Authentication successful!\n\n")
+			return authToken, nil
+
+		}
+
+		return "", nil
```

Comment:
Mostly a nit (since it looks like this is unreachable) but if we somehow got here, we probably shouldn't return nil for the error (or at least a comment to the effect that it's unreachable so that a cursory examination of the code won't make it appear that this is a valid exit :thinking: